src/mouse_events.cpp

```diff
@@ -1484,8 +1515,50 @@ void mouse_handler::show_attack_options(const unit_map::const_iterator& u)
 	const team& cur_team = current_team();
 	const team& u_team = pc_.get_teams()[u->side() - 1];
 
-	// Check each adjacent hex.
-	for(const map_location& loc : get_adjacent_tiles(u->get_location())) {
+
+	// convert to cubical -> enumerate tiles -> convert back
+	// see https://www.redblobgames.com/grids/hexagons/ for detailed description
+
+
+	// convert to cubical coordinates
+	// cx == a.x, cy is not needed
+	const auto& attacks = u->attacks();
+	std::set<int> attackable_distances;
+
+    for (const auto& attack : attacks) {
+        for (int i = attack.min_range(); i <= attack.max_range(); ++i) {
+            attackable_distances.insert(i);
+        }
+    }
+
+	// for optimalization, I think iterating over ranges here, would be better
+
+	int min = *attackable_distances.begin();
+	int n = *std::prev(attackable_distances.end());
+	const map_location& a = u->get_location();
+
+	int cz = a.y - (a.x - (a.x&1)) / 2;
```

Comment:
As with the other PR, let's use `floor(a.x / 2.0)` here.

Reply:
This is generally used code man, this is the current best way.
...
Do we really have to do this?

Reply:
I think it's clearer if code expresses is intent rather than using obscure bitwise operations. Yes, even if "everyone else" uses obscure bitwise operations.

Based on the link you put in the code a few lines up, there's the suggestion of using `abs(x)%2` instead of `x&1`, which would be slightly better, but `(x - x%2) / 2` still looks weird (and divides _twice_!), so I think the floor is still preferred.

Although, it would be even better if the logic is extracted out into a function, probably a member function of `map_location`. It could be called `to_cubic` or similar and return a `tuple<int,int,int>` or a new custom struct with three members `q, r, s`.

Reply:
Bruh